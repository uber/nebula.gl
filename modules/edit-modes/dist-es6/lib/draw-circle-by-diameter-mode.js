"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawCircleByDiameterMode = void 0;

var _circle = _interopRequireDefault(require("@turf/circle"));

var _distance = _interopRequireDefault(require("@turf/distance"));

var _geojsonEditMode = require("./geojson-edit-mode");

var _twoClickPolygonMode = require("./two-click-polygon-mode");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var DrawCircleByDiameterMode = /*#__PURE__*/function (_TwoClickPolygonMode) {
  _inherits(DrawCircleByDiameterMode, _TwoClickPolygonMode);

  var _super = _createSuper(DrawCircleByDiameterMode);

  function DrawCircleByDiameterMode() {
    _classCallCheck(this, DrawCircleByDiameterMode);

    return _super.apply(this, arguments);
  }

  _createClass(DrawCircleByDiameterMode, [{
    key: "getTwoClickPolygon",
    value: function getTwoClickPolygon(coord1, coord2, modeConfig) {
      // Default turf value for circle is 64
      var _ref = modeConfig || {},
          _ref$steps = _ref.steps,
          steps = _ref$steps === void 0 ? 64 : _ref$steps;

      var options = {
        steps: steps
      };

      if (steps < 4) {
        console.warn("Minimum steps to draw a circle is 4 "); // eslint-disable-line no-console,no-undef

        options.steps = 4;
      }

      var centerCoordinates = (0, _geojsonEditMode.getIntermediatePosition)(coord1, coord2);
      var radius = Math.max((0, _distance["default"])(coord1, centerCoordinates), 0.001);
      var geometry = (0, _circle["default"])(centerCoordinates, radius, options);
      geometry.properties = geometry.properties || {};
      geometry.properties.shape = 'Circle';
      return geometry;
    }
  }]);

  return DrawCircleByDiameterMode;
}(_twoClickPolygonMode.TwoClickPolygonMode);

exports.DrawCircleByDiameterMode = DrawCircleByDiameterMode;
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uL3NyYy9saWIvZHJhdy1jaXJjbGUtYnktZGlhbWV0ZXItbW9kZS50cyJdLCJuYW1lcyI6WyJEcmF3Q2lyY2xlQnlEaWFtZXRlck1vZGUiLCJjb29yZDEiLCJjb29yZDIiLCJtb2RlQ29uZmlnIiwic3RlcHMiLCJvcHRpb25zIiwiY29uc29sZSIsIndhcm4iLCJjZW50ZXJDb29yZGluYXRlcyIsInJhZGl1cyIsIk1hdGgiLCJtYXgiLCJnZW9tZXRyeSIsInByb3BlcnRpZXMiLCJzaGFwZSIsIlR3b0NsaWNrUG9seWdvbk1vZGUiXSwibWFwcGluZ3MiOiI7Ozs7Ozs7OztBQUFBOztBQUNBOztBQUVBOztBQUNBOzs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7SUFFYUEsd0I7Ozs7Ozs7Ozs7Ozs7dUNBQ1FDLE0sRUFBa0JDLE0sRUFBa0JDLFUsRUFBcUM7QUFDMUY7QUFEMEYsaUJBRW5FQSxVQUFVLElBQUksRUFGcUQ7QUFBQSw0QkFFbEZDLEtBRmtGO0FBQUEsVUFFbEZBLEtBRmtGLDJCQUUxRSxFQUYwRTs7QUFHMUYsVUFBTUMsT0FBTyxHQUFHO0FBQUVELFFBQUFBLEtBQUssRUFBTEE7QUFBRixPQUFoQjs7QUFFQSxVQUFJQSxLQUFLLEdBQUcsQ0FBWixFQUFlO0FBQ2JFLFFBQUFBLE9BQU8sQ0FBQ0MsSUFBUix5Q0FEYSxDQUN5Qzs7QUFDdERGLFFBQUFBLE9BQU8sQ0FBQ0QsS0FBUixHQUFnQixDQUFoQjtBQUNEOztBQUVELFVBQU1JLGlCQUFpQixHQUFHLDhDQUF3QlAsTUFBeEIsRUFBZ0NDLE1BQWhDLENBQTFCO0FBQ0EsVUFBTU8sTUFBTSxHQUFHQyxJQUFJLENBQUNDLEdBQUwsQ0FBUywwQkFBU1YsTUFBVCxFQUFpQk8saUJBQWpCLENBQVQsRUFBOEMsS0FBOUMsQ0FBZjtBQUVBLFVBQU1JLFFBQVEsR0FBRyx3QkFBT0osaUJBQVAsRUFBMEJDLE1BQTFCLEVBQWtDSixPQUFsQyxDQUFqQjtBQUVBTyxNQUFBQSxRQUFRLENBQUNDLFVBQVQsR0FBc0JELFFBQVEsQ0FBQ0MsVUFBVCxJQUF1QixFQUE3QztBQUNBRCxNQUFBQSxRQUFRLENBQUNDLFVBQVQsQ0FBb0JDLEtBQXBCLEdBQTRCLFFBQTVCO0FBRUEsYUFBT0YsUUFBUDtBQUNEOzs7O0VBcEIyQ0csd0MiLCJzb3VyY2VzQ29udGVudCI6WyJpbXBvcnQgY2lyY2xlIGZyb20gJ0B0dXJmL2NpcmNsZSc7XG5pbXBvcnQgZGlzdGFuY2UgZnJvbSAnQHR1cmYvZGlzdGFuY2UnO1xuaW1wb3J0IHsgUG9zaXRpb24sIFBvbHlnb24sIEZlYXR1cmVPZiB9IGZyb20gJy4uL2dlb2pzb24tdHlwZXMnO1xuaW1wb3J0IHsgZ2V0SW50ZXJtZWRpYXRlUG9zaXRpb24gfSBmcm9tICcuL2dlb2pzb24tZWRpdC1tb2RlJztcbmltcG9ydCB7IFR3b0NsaWNrUG9seWdvbk1vZGUgfSBmcm9tICcuL3R3by1jbGljay1wb2x5Z29uLW1vZGUnO1xuXG5leHBvcnQgY2xhc3MgRHJhd0NpcmNsZUJ5RGlhbWV0ZXJNb2RlIGV4dGVuZHMgVHdvQ2xpY2tQb2x5Z29uTW9kZSB7XG4gIGdldFR3b0NsaWNrUG9seWdvbihjb29yZDE6IFBvc2l0aW9uLCBjb29yZDI6IFBvc2l0aW9uLCBtb2RlQ29uZmlnOiBhbnkpOiBGZWF0dXJlT2Y8UG9seWdvbj4ge1xuICAgIC8vIERlZmF1bHQgdHVyZiB2YWx1ZSBmb3IgY2lyY2xlIGlzIDY0XG4gICAgY29uc3QgeyBzdGVwcyA9IDY0IH0gPSBtb2RlQ29uZmlnIHx8IHt9O1xuICAgIGNvbnN0IG9wdGlvbnMgPSB7IHN0ZXBzIH07XG5cbiAgICBpZiAoc3RlcHMgPCA0KSB7XG4gICAgICBjb25zb2xlLndhcm4oYE1pbmltdW0gc3RlcHMgdG8gZHJhdyBhIGNpcmNsZSBpcyA0IGApOyAvLyBlc2xpbnQtZGlzYWJsZS1saW5lIG5vLWNvbnNvbGUsbm8tdW5kZWZcbiAgICAgIG9wdGlvbnMuc3RlcHMgPSA0O1xuICAgIH1cblxuICAgIGNvbnN0IGNlbnRlckNvb3JkaW5hdGVzID0gZ2V0SW50ZXJtZWRpYXRlUG9zaXRpb24oY29vcmQxLCBjb29yZDIpO1xuICAgIGNvbnN0IHJhZGl1cyA9IE1hdGgubWF4KGRpc3RhbmNlKGNvb3JkMSwgY2VudGVyQ29vcmRpbmF0ZXMpLCAwLjAwMSk7XG5cbiAgICBjb25zdCBnZW9tZXRyeSA9IGNpcmNsZShjZW50ZXJDb29yZGluYXRlcywgcmFkaXVzLCBvcHRpb25zKTtcblxuICAgIGdlb21ldHJ5LnByb3BlcnRpZXMgPSBnZW9tZXRyeS5wcm9wZXJ0aWVzIHx8IHt9O1xuICAgIGdlb21ldHJ5LnByb3BlcnRpZXMuc2hhcGUgPSAnQ2lyY2xlJztcblxuICAgIHJldHVybiBnZW9tZXRyeTtcbiAgfVxufVxuIl19