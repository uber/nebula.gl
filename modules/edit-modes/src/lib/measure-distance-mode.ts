// @ts-ignore
import turfDistance from '@turf/distance';
// @ts-ignore
import memoize from '../memoize';
import {
  ClickEvent,
  PointerMoveEvent,
  Tooltip,
  ModeProps,
  GuideFeatureCollection,
  EditHandleFeature,
} from '../types';
import { FeatureCollection, Position } from '../geojson-types';
import { GeoJsonEditMode } from './geojson-edit-mode';

const DEFAULT_TOOLTIPS = [];

export class MeasureDistanceMode extends GeoJsonEditMode {
  startingPoint: Readonly<EditHandleFeature> | null | undefined = null;
  endingPoint: Readonly<EditHandleFeature> | null | undefined = null;
  endingPointLocked = false;

  _setEndingPoint(mapCoords: Position) {
    this.endingPoint = {
      type: 'Feature',
      properties: {
        guideType: 'editHandle',
        editHandleType: 'existing',
        featureIndex: -1,
        positionIndexes: [],
      },
      geometry: {
        type: 'Point',
        coordinates: mapCoords,
      },
    };
  }

  _getTooltips = memoize(({ modeConfig, startingPoint, endingPoint }) => {
    let tooltips = DEFAULT_TOOLTIPS;

    if (startingPoint && endingPoint) {
      const { formatTooltip, turfOptions, measurementCallback } = modeConfig || {};
      const units = (turfOptions && turfOptions.units) || 'kilometers';

      const distance = turfDistance(startingPoint, endingPoint, turfOptions);

      let text;
      if (formatTooltip) {
        text = formatTooltip(distance);
      } else {
        // By default, round to 2 decimal places and append units
        text = `${parseFloat(distance).toFixed(2)} ${units}`;
      }

      if (measurementCallback) {
        measurementCallback(distance);
      }

      tooltips = [
        {
          position: endingPoint.geometry.coordinates,
          text,
        },
      ];
    }

    return tooltips;
  });

  handleClick(event: ClickEvent, props: ModeProps<FeatureCollection>): void {
    if (!this.startingPoint || this.endingPointLocked) {
      this.startingPoint = {
        type: 'Feature',
        properties: {
          guideType: 'editHandle',
          editHandleType: 'existing',
          featureIndex: -1,
          positionIndexes: [],
        },
        geometry: {
          type: 'Point',
          coordinates: event.mapCoords,
        },
      };
      this.endingPoint = null;
      this.endingPointLocked = false;
    } else if (this.startingPoint) {
      this._setEndingPoint(event.mapCoords);
      this.endingPointLocked = true;
    }
  }

  // Called when the pointer moved, regardless of whether the pointer is down, up, and whether something was picked
  handlePointerMove(event: PointerMoveEvent, props: ModeProps<FeatureCollection>): void {
    if (this.startingPoint && !this.endingPointLocked) {
      this._setEndingPoint(event.mapCoords);
    }

    props.onUpdateCursor('cell');
  }

  // Return features that can be used as a guide for editing the data
  getGuides(props: ModeProps<FeatureCollection>): GuideFeatureCollection {
    const guides: GuideFeatureCollection = { type: 'FeatureCollection', features: [] };
    const { features } = guides;

    if (this.startingPoint) {
      features.push(this.startingPoint);
    }
    if (this.endingPoint) {
      features.push(this.endingPoint);
    }
    if (this.startingPoint && this.endingPoint) {
      features.push({
        type: 'Feature',
        properties: { guideType: 'tentative' },
        geometry: {
          type: 'LineString',
          coordinates: [
            this.startingPoint.geometry.coordinates,
            this.endingPoint.geometry.coordinates,
          ],
        },
      });
    }
    return guides;
  }

  getTooltips(props: ModeProps<FeatureCollection>): Tooltip[] {
    return this._getTooltips({
      modeConfig: props.modeConfig,
      startingPoint: this.startingPoint,
      endingPoint: this.endingPoint,
    });
  }
}
