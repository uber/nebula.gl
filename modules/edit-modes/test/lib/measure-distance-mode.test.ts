/* eslint-env jest */

import { MeasureDistanceMode } from '../../src/lib/measure-distance-mode';
import {
  createFeatureCollectionProps,
  createClickEvent,
  createPointerMoveEvent,
  createKeyboardEvent,
} from '../test-utils';

const expectToBeCloseToArray = (actual, expected) => {
  expect(actual.length).toBe(expected.length);
  actual.forEach((x, index) => expect(x).toBeCloseTo(expected[index], 1));
};

describe('move without click', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handlePointerMove(createPointerMoveEvent(), createFeatureCollectionProps());
  });

  it('guides are empty', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides).toEqual({ type: 'FeatureCollection', features: [] });
  });

  it('tooltips are empty', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toEqual([]);
  });
});

describe('one click', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
  });

  it('guides are a single point', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides).toMatchSnapshot();
  });

  it('tooltips are 0.00', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips[0].text).toContain('0.00');
  });
});

describe('one click - centerTooltipsOnLine = true', () => {
  let mode;
  const props = { modeConfig: { centerTooltipsOnLine: true } };
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps(props));
  });

  it('tooltip is placed at [1,2]', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps(props));
    expect(tooltips[0].position).toEqual([1, 2]);
  });
});

describe('one click + pointer move', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handlePointerMove(createPointerMoveEvent([3, 4]), createFeatureCollectionProps());
  });

  it('guides are two points + line string', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides).toMatchSnapshot();
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });
});

describe('one click + pointer move - centerTooltipsOnLine = true', () => {
  let mode;
  const props = { modeConfig: { centerTooltipsOnLine: true } };
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps(props));
    mode.handlePointerMove(createPointerMoveEvent([1, 4]), createFeatureCollectionProps(props));
  });

  it('tooltips are on center of their respective line', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps(props));
    expect(tooltips[0].position).toEqual([1, 3]);
  });
});

describe('two clicks', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([3, 4]), createFeatureCollectionProps());
  });

  it('guides are two points + line string', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides).toMatchSnapshot();
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });

  it('can measure kilometers', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips[0].text).toContain('kilometers');
  });

  it('can measure miles', () => {
    const tooltips = mode.getTooltips(
      createFeatureCollectionProps({ modeConfig: { turfOptions: { units: 'miles' } } })
    );
    expect(tooltips[tooltips.length - 1].text).toContain('miles');
  });

  it('can format distance', () => {
    const tooltips = mode.getTooltips(
      createFeatureCollectionProps({ modeConfig: { formatTooltip: String } })
    );
    expect(tooltips[tooltips.length - 1].text).toEqual('314.28368918020476');
  });
});

describe('two clicks - centerTooltipsOnLine = true', () => {
  let mode;
  const props = { modeConfig: { centerTooltipsOnLine: true } };

  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps(props));
    mode.handleClick(createClickEvent([1, 4]), createFeatureCollectionProps(props));
  });

  it('tooltips are on center of their respective line', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps(props));
    expect(tooltips[0].position).toEqual([1, 3]);
    expect(tooltips[1].position).toEqual([1, 4]);
  });
});

describe('two clicks + pointer move', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([3, 4]), createFeatureCollectionProps());
    mode.handlePointerMove(createPointerMoveEvent([4, 5]), createFeatureCollectionProps());
  });

  it('ending point is clicked point not hovered point', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides.features[2].geometry.coordinates).toEqual([3, 4]);
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });
});

describe('two clicks + pointer move - centerTooltipsOnLine = true', () => {
  let mode;
  const props = { modeConfig: { centerTooltipsOnLine: true } };

  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps(props));
    mode.handleClick(createClickEvent([1, 4]), createFeatureCollectionProps(props));
    mode.handlePointerMove(createPointerMoveEvent([5, 4]), createFeatureCollectionProps(props));
  });

  it('tooltips are on center of their respective line', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps(props));
    expect(tooltips[0].position).toEqual([1, 3]);
    expectToBeCloseToArray(tooltips[1].position, [3, 4]);
  });
});

describe('three clicks + pointer move', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([3, 4]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([4, 5]), createFeatureCollectionProps());
    mode.handlePointerMove(createPointerMoveEvent([6, 7]), createFeatureCollectionProps());
  });

  it('first feature is a tentative line that contains 4 points', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides.features[0].properties.guideType).toEqual('tentative');
    expect(guides.features[0].geometry.type).toEqual('LineString');
    expect(guides.features[0].geometry.coordinates.length).toEqual(4);
    expect(guides.features[0].geometry.coordinates[3]).toEqual([6, 7]);
  });

  it('Second feature is a editHandle point with coordinates [1,2]', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides.features[1].properties.guideType).toEqual('editHandle');
    expect(guides.features[1].geometry.type).toEqual('Point');
    expect(guides.features[1].geometry.coordinates).toEqual([1, 2]);
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });
});

describe('three clicks + pointer move - centerTooltipsOnLine = true', () => {
  let mode;
  const props = { modeConfig: { centerTooltipsOnLine: true } };

  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps(props));
    mode.handleClick(createClickEvent([1, 4]), createFeatureCollectionProps(props));
    mode.handleClick(createClickEvent([5, 4]), createFeatureCollectionProps(props));
    mode.handlePointerMove(createPointerMoveEvent([11, 8]), createFeatureCollectionProps(props));
  });

  it('tooltips are on center of their respective line', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps(props));
    expect(tooltips[0].position).toEqual([1, 3]);
    expectToBeCloseToArray(tooltips[1].position, [3, 4]);
    expectToBeCloseToArray(tooltips[2].position, [8, 6]);
  });
});

describe('three clicks + pointer move + press Escape', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([3, 4]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([4, 5]), createFeatureCollectionProps());
    mode.handlePointerMove(createPointerMoveEvent([6, 7]), createFeatureCollectionProps());
    mode.handleKeyUp(createKeyboardEvent('Escape'), createFeatureCollectionProps());
  });

  it('first feature is a tentative line that contains 3 points', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides.features[0].properties.guideType).toEqual('tentative');
    expect(guides.features[0].geometry.type).toEqual('LineString');
    expect(guides.features[0].geometry.coordinates.length).toEqual(3);
    expect(guides.features[0].geometry.coordinates[2]).toEqual([4, 5]);
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });
});

describe('three clicks + pointer move + press Enter', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([3, 4]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([4, 5]), createFeatureCollectionProps());
    mode.handlePointerMove(createPointerMoveEvent([6, 7]), createFeatureCollectionProps());
    mode.handleKeyUp(createKeyboardEvent('Enter'), createFeatureCollectionProps());
  });

  it('first feature is a tentative line that contains 3 points', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides.features[0].properties.guideType).toEqual('tentative');
    expect(guides.features[0].geometry.type).toEqual('LineString');
    expect(guides.features[0].geometry.coordinates.length).toEqual(4);
    expect(guides.features[0].geometry.coordinates[3]).toEqual([6, 7]);
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });
});
