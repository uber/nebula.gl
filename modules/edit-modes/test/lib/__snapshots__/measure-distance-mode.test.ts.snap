// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one click + pointer move guides are two points + line string 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            1,
            2,
          ],
          Array [
            3,
            4,
          ],
        ],
        "type": "LineString",
      },
      "properties": Object {
        "guideType": "tentative",
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`one click + pointer move tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
]
`;

exports[`one click guides are a single point 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            1,
            2,
          ],
          Array [
            1,
            2,
          ],
        ],
        "type": "LineString",
      },
      "properties": Object {
        "guideType": "tentative",
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`three clicks + pointer move + press Enter tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
  Object {
    "position": Array [
      4,
      5,
    ],
    "text": "471.29 kilometers",
  },
  Object {
    "position": Array [
      6,
      7,
    ],
    "text": "784.93 kilometers",
  },
]
`;

exports[`three clicks + pointer move + press Escape tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
  Object {
    "position": Array [
      4,
      5,
    ],
    "text": "471.29 kilometers",
  },
]
`;

exports[`three clicks + pointer move tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
  Object {
    "position": Array [
      4,
      5,
    ],
    "text": "471.29 kilometers",
  },
  Object {
    "position": Array [
      6,
      7,
    ],
    "text": "784.93 kilometers",
  },
]
`;

exports[`two clicks + pointer move tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
  Object {
    "position": Array [
      4,
      5,
    ],
    "text": "471.29 kilometers",
  },
]
`;

exports[`two clicks guides are two points + line string 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            1,
            2,
          ],
          Array [
            3,
            4,
          ],
          Array [
            3,
            4,
          ],
        ],
        "type": "LineString",
      },
      "properties": Object {
        "guideType": "tentative",
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`two clicks tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
]
`;
